### PR TITLE
Moab cmake

### PR DESCRIPTION
**Pull request by gonuke**
This changes the build process for MOAB in the docker image to use CMake instead of Autotools.

This build system is more robust across different platforms and will be where ongoing improvements are focused.

depends on PR #684

### CODE REVIEW

**Comment by pshriwise** [comment]
In the past DAGMC has required that both static and shared libraries are included in the MOAB install but to my knowledge CMake can only build and install one set of those at a time. From these changes it looks like we'll no longer have a set of static libraries in the Travis docker image. (It seems the CMake default for MOAB is to build shared object libraries.)

**Comment by gonuke** [comment]
Good point.  I've been advocating a move to shared libraries only but I can't recall if there is a real reason to keep static libraries?  @ljacobson64 @makeclean ?

**Comment by gonuke** [comment]
This might break CI since it tries to build both static and shared versions...???

**Comment by pshriwise** [comment]
Yeah I'd guess that it would. An additional build and install with `BUILD_SHARED_LIBS=OFF` has done the trick for me in the past.

**Comment by gonuke** [comment]
I added a static lib build for now - we can push off the debate for static + shared to another time....

**Comment by bam241** [comment]
LGTM! 
Did someone test ro rebuild the docker image using this ?

**Comment by gonuke** [comment]
I haven’t tested a new image... I suppose the best plan is to test the CI building & test process on a new image built with this

**Comment by bam241** [comment]
I'll do that later today

**Comment by bam241** [comment]
ubuntu 16.04 seems to be working.
testing 18.04 right now.

**Comment by bam241** [comment]
having issue with 18.04, all the test failed with:
```
est project /root/build/gcc/DAGMC-moab-5.1.0-hdf5-1.10.4-shared/bld
      Start  1: dagmc_unit_tests
 1/24 Test  #1: dagmc_unit_tests ........................***Failed    0.00 sec
/root/build/gcc/DAGMC-moab-5.1.0-hdf5-1.10.4-shared/bld/src/dagmc/tests/dagmc_unit_tests: error while loading shared libraries: libhdf5.so.103: cannot open shared object file: No such file or directory
```

**Comment by pshriwise** [comment]
If it isn't too much trouble, could we add 19.10 into the mix as well? I vaguely recall having some problems that were unique to building MOAB with CMake in that version...

**Comment by gonuke** [comment]
Shouldn't we leap to 20.04 then?

**Comment by ljacobson64** [comment]
I think we should only have LTS releases (16.04, 18.04, 20.04) in CI

**Comment by pshriwise** [comment]
Mmm yeah that would make more sense.

**Comment by gonuke** [comment]
So back to the original error....  

It seems that the custom HDF5 shared lib is not being found, so it's not in the library path or rpath is not right.... 

Also, do we only test in CI against the custom HDF5 and never against the system HDF5?  Or am I reading something wrong...

**Comment by ljacobson64** [comment]
We only test with the custom HDF5.

https://github.com/svalinn/DAGMC/issues/606

**Comment by bam241** [comment]
@gonuke I still have the issue, but not on build, on test only.

I am wondering is this could be related to the issue Helen mention this morning ?!

**Comment by pshriwise** [comment]
@bam241 did you try running the installed test executeables? Same issue?

**Comment by bam241** [comment]
@pshriwise I did not try, I use a script that mimics what CI is supposed to do.
(I did not investigate this to much either :))

**Comment by gonuke** [comment]
I suspect a combination of things here, but don't have time to test:

- autotools is setting the rpath when it builds MOAB so that it knows where to find HDF5 and CMake is not
- the LD_LIBRARY_PATH is being cleared by ctest (per Helen's discovery in the CMakeLists.txt file)

This would only fail when running ctest after building MOAB with CMake.

**Comment by ljacobson64** [comment]
If the problem is that the CMake build isn't adding the rpath, this can be fixed without needing to change MOAB at all by adding `-DCMAKE_INSTALL_RPATH=${hdf5_install_dir}/lib:${moab_install_dir}/lib` to the MOAB cmake command

**Comment by gonuke** [comment]
I added the RPATH line that @ljacobson64 suggested - maybe that will help

**Comment by ljacobson64** [comment]
Looks like that fix worked -- the tests are passing on Travis

**Comment by gonuke** [comment]
CI is not testing this since this is just a change in the Dockerfiles that build the images

**Comment by ljacobson64** [comment]
Right, never mind

**Comment by shimwell** [comment]
Thanks for making this, I think it will be most useful for an open OpenMC issue https://github.com/openmc-dev/openmc/issues/1709 where we are trying to make the automated OpenMC docker image optionally include DAGMC

**Comment by bam241** [comment]
what is the status there ?

**Comment by gonuke** [comment]
This has been rebased against changes proposed in #735

**Review by ljacobson64 — APPROVED** [review]
This LGTM, assuming the tests pass after the docker image is rebuilt

**Comment by gonuke** [comment]
I'm rebuilding docker images now (1/2 done) and will test later and update

**Comment by gonuke** [comment]
All tests passed on new docker images.  This CI should be operating on those same docker images now.

**Comment by gonuke** [comment]
I guess they didn't pass :(

File: CI/docker/build_moab.sh
Comment (ljacobson64) [inline CI/docker/build_moab.sh]: I think that replacing
```
-DCMAKE_INSTALL_RPATH=${hdf5_install_dir}:${moab_install_dir}
```
with
```
-DCMAKE_INSTALL_RPATH=${hdf5_install_dir}/lib:${moab_install_dir}/lib
```
here will fix the current CI issue.

File: CI/docker/build_moab.sh
Comment (ljacobson64) [inline CI/docker/build_moab.sh]: Same comment here.

**Comment by gonuke** [comment]
I'll try that a little later today

File: CI/docker/build_moab.sh
Comment (ljacobson64) [reply CI/docker/build_moab.sh]: https://github.com/svalinn/DAGMC/pull/687#issuecomment-656489283

Comment (gonuke) [reply CI/docker/build_moab.sh]: It looks like I might have dropped the `/lib` when adapting this in  the last update

**Comment by gonuke** [comment]
Rebuilding docker images this morning (hopefully) to push this afternoon for final testing :crossed_fingers:

**Comment by gonuke** [comment]
Relaunched with new docker images.  Everything passed except the elusive overlap_check tests that appear to be a MOAB problem?

**Comment by ljacobson64** [comment]
Weird... and the tests pass with clang but not with gcc

**Comment by gonuke** [comment]
This problem has been popping up more & more often... ☹️

**Comment by bam241** [comment]
is that not a problem related to MOAB ?! (I vaguely remember @pshriwise tell me this....)

**Comment by gonuke** [comment]
Yes - I think he's working on it in MOAB

**Comment by ljacobson64** [comment]
Seems odd that switching moab's build system from autotools to cmake would cause this issue to pop up

**Comment by bam241** [comment]
I believe this problem predate the proposed change to switch to MOAB build to CMake.

**Comment by bam241** [comment]
the root of the error might be explained there: https://stackoverflow.com/questions/30887143/make-j-8-g-internal-compiler-error-killed-program-cc1plus

**Comment by gonuke** [comment]
Should we combine this with #726 while we're waiting?

**Comment by pshriwise** [comment]
It looks like this now needs to be rebased to ensure we're using the updated docker images with a recent version of MOAB. Is that right?

**Comment by pshriwise** [comment]
@ljacobson64 I know it's been a while, but is this looking good to you still?

**Comment by gonuke** [comment]
I suppose someone should rebuild the CI images and confirm that they work

**Comment by gonuke** [comment]
Or we can wait for the MOAB release for this one....

**Comment by gonuke** [comment]
This was replaced by #771 in our new build infrastructure.

**Pull request closed without merging**
